src/backend/distributed/deparser/deparse_role_stmts.c

```diff
@@ -427,28 +437,100 @@ AppendGrantWithAdminOption(StringInfo buf, GrantRoleStmt *stmt)
-			if (strcmp(opt->defname, "admin") == 0 &&
-				parse_bool(optval, &admin_option) && admin_option)
+			bool option_value = false;
+			if (parse_bool(optval, &option_value))
 			{
-				appendStringInfo(buf, " WITH ADMIN OPTION");
-				break;
+				opt_count++;
+				char *prefix = opt_count > 1 ? ", " : " WITH ";
+				if (strcmp(opt->defname, "inherit") == 0)
+				{
+					appendStringInfo(buf, "%s INHERIT %s", prefix, option_value ? "TRUE" :
+									 "FALSE");
+				}
+				else if (strcmp(opt->defname, "admin") == 0 && option_value)
+				{
+					appendStringInfo(buf, "%s ADMIN OPTION", prefix);
+				}
+				else if (strcmp(opt->defname, "set") == 0)
+				{
+					appendStringInfo(buf, "%s SET %s", prefix, option_value ? "TRUE" :
+									 "FALSE");
+				}
 			}
 		}
-#else
+		#else
 		if (stmt->admin_opt)
 		{
 			appendStringInfo(buf, " WITH ADMIN OPTION");
 		}
-#endif
+		#endif
 	}
 }
 
 
+/* #if PG_VERSION_NUM >= PG_VERSION_16 */
+/*      int opt_count = 0; */
+/*      DefElem *opt = NULL; */
+/*      foreach_ptr(opt, stmt->opt) */
+/*      { */
+/*          if (strcmp(opt->defname, "inherit") == 0) */
+/*          { */
+/*              bool inherit_option = false; */
+/*              char *optval = defGetString(opt); */
+/*              if (parse_bool(optval, &inherit_option)) */
+/*              { */
+/*                  opt_count++; */
+/*                  if (!inherit_option) */
+/*                  { */
+/*                      appendStringInfo(buf, opt_count > 1 ? ", INHERIT FALSE" : "WITH INHERIT FALSE"); */
+/*                  } */
+/*                  else */
+/*                  { */
+/*                      appendStringInfo(buf, opt_count > 1 ? ", INHERIT TRUE" : "WITH INHERIT TRUE"); */
+/*                  } */
+/*              } */
+/*          } */
+/*          else if (strcmp(opt->defname, "admin") == 0) */
+/*          { */
+/*              bool admin_option = false; */
+/*              char *optval = defGetString(opt); */
+/*              if (parse_bool(optval, &admin_option) && admin_option) */
+/*              { */
+/*                  appendStringInfo(buf, ", ADMIN OPTION"); */
+/*              } */
+/*          } */
+/*          else if (strcmp(opt->defname, "set") == 0) */
+/*          { */
+/*              bool set_option = false; */
+/*              char *optval = defGetString(opt); */
+/*              if (parse_bool(optval, &set_option)) */
+/*              { */
+/*                  if (!set_option) */
+/*                  { */
+/*                      appendStringInfo(buf, ", SET FALSE"); */
+/*                  } */
+/*                  else */
+/*                  { */
+/*                      appendStringInfo(buf, ", SET TRUE"); */
+/*                  } */
+/*              } */
+/*          } */
+/*      } */
+/* #else */
+/*      if (stmt->admin_opt) */
+/*      { */
+/*          appendStringInfo(buf, " WITH ADMIN OPTION"); */
+/*      } */
+/* #endif */
+/*  } */
+/* } */
+
+
```

Comment:
Let's remove this :)

Reply:
oops ... Totally forgot about this. Must have been a late night. thx